### PR TITLE
Enable dequant+matmul 8bit path for Intel CPU and XPU

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @Titus-von-Koeller @matthewdouglas . This feature enables dequant 8bit weight and using float matmul. It speed-up the lora finetune for 3x on XPU and 2x on CPU by the [lora finetune script](https://github.com/huggingface/peft/blob/main/examples/olora_finetuning/olora_finetuning.py) on llama3-8b by the command `python olora_finetuning.py --base_model alokabhishek/Meta-Llama-3-8B-Instruct-bnb-8bit --init_lora_weights gaussian --seed 42 --torch_dtype bfloat16 --device_map cpu`. 

All tests in transformers have been passed, please review this PR. Thanks!

### CODE REVIEW

**Comment by matthewdouglas** [comment]
On the mainline branch with the int8 refactoring that was done for v0.45.0, `Linear8bitLt` was simplified a bit. One of the optimizations made there was to only do the row-wise quantization for inference, and still do the "double quant" with row/col for training.

 But with that said the decomposition into separate int8 and fp16 matmuls has some overhead and I assume this is where the unsafe operations come from, particularly when threshold!=0.

Happy to merge; we can revisit doing int8 computation in the future if needed.

**Review by matthewdouglas — APPROVED** [review]
